File: fixtures/nodejs/j03/fixture.js
// Shared counter with race condition — no mutex/lock
// Multiple concurrent async operations increment without coordination

let sharedCounter = 0;

async function incrementCounter(times) {
  for (let i = 0; i < times; i++) {
    // Race condition: read-modify-write is not atomic
    const current = sharedCounter;
    await new Promise((resolve) => setImmediate(resolve)); // yield to event loop
    sharedCounter = current + 1;
  }
}

async function getCounter() {
  return sharedCounter;
}

async function resetCounter() {
  sharedCounter = 0;
}

module.exports = { incrementCounter, getCounter, resetCounter };

File: fixtures/dotnet/j03/Fixture.cs
// Shared counter with race condition -- no lock/Interlocked
// Multiple concurrent async operations increment without coordination

public class Counter
{
    private int _sharedCounter = 0;

    public async Task IncrementCounter(int times)
    {
        for (int i = 0; i < times; i++)
        {
            // Race condition: read-modify-write is not atomic
            int current = _sharedCounter;
            await Task.Delay(0); // yield to allow other tasks to interleave
            _sharedCounter = current + 1;
        }
    }

    public Task<int> GetCounter()
    {
        return Task.FromResult(_sharedCounter);
    }

    public Task ResetCounter()
    {
        _sharedCounter = 0;
        return Task.CompletedTask;
    }
}
